### PR TITLE
ARROW-764: [C++] Improves performance of CopyBitmap and adds benchmarks

### PR DESCRIPTION
**Pull request by seibs**
I took a swing at improving the CopyBitmap performance (benchmarks below).  I'm a C/C++ novice, so I thought I'd get some feedback before I went too much further.

**Starting Point**
```
Run on (4 X 2208 MHz CPU s)
12/13/17 21:15:18
Benchmark                                        Time           CPU Iterations
------------------------------------------------------------------------------
BM_CopyBitmap/97.6563k/0/min_time:1.000       4779 us       4758 us        289   20.0445MB/s
BM_CopyBitmap/976.563k/0/min_time:1.000      47740 us      47476 us         26   20.0875MB/s
BM_CopyBitmap/97.6563k/4/min_time:1.000       4858 us       4866 us        289   19.5991MB/s
BM_CopyBitmap/976.563k/4/min_time:1.000      48117 us      47953 us         29   19.8879MB/s
```

**Using stanford bithacks for SetBitTo**
```
Run on (4 X 2208 MHz CPU s)
12/13/17 21:22:05
Benchmark                                        Time           CPU Iterations
------------------------------------------------------------------------------
BM_CopyBitmap/97.6563k/0/min_time:1.000       1647 us       1649 us        815   57.8415MB/s
BM_CopyBitmap/976.563k/0/min_time:1.000      16368 us      16397 us         81   58.1629MB/s
BM_CopyBitmap/97.6563k/4/min_time:1.000       1599 us       1610 us        815   59.2186MB/s
BM_CopyBitmap/976.563k/4/min_time:1.000      16026 us      16011 us         81   59.5644MB/s
```

**memcpy + shifting**
*This solution provides varying performance depending on whether or not the bit offset is a multiple of 8*
```
Run on (4 X 2208 MHz CPU s)
12/13/17 21:23:44
Benchmark                                        Time           CPU Iterations
------------------------------------------------------------------------------
BM_CopyBitmap/97.6563k/0/min_time:1.000          5 us          5 us     280000   18.9651GB/s
BM_CopyBitmap/976.563k/0/min_time:1.000         62 us         61 us      22400   15.1721GB/s
BM_CopyBitmap/97.6563k/4/min_time:1.000        171 us        170 us       6892   560.872MB/s
BM_CopyBitmap/976.563k/4/min_time:1.000       1639 us       1639 us        896   581.782MB/s
```

### CODE REVIEW

File: cpp/src/arrow/compute/compute-benchmark.cc [hunk 1]
```diff
@@ -191,7 +191,7 @@ static void BM_UniqueString100bytes(benchmark::State& state) {
 BENCHMARK(BM_BuildDictionary)->MinTime(1.0)->Unit(benchmark::kMicrosecond);
 BENCHMARK(BM_BuildStringDictionary)->MinTime(1.0)->Unit(benchmark::kMicrosecond);
 
-constexpr int64_t kHashBenchmarkLength = 1 << 24;
+constexpr int kHashBenchmarkLength = 1 << 24;
```
Comment (xhochy) [inline cpp/src/arrow/compute/compute-benchmark.cc]: Can you keep this an `int64_t`?

File: cpp/src/arrow/util/CMakeLists.txt [hunk 2]
```diff
@@ -57,11 +57,16 @@ if (ARROW_BUILD_BENCHMARKS)
...
       benchmark
     )
+  elseif(MSVC)
+    target_link_libraries(arrow_benchmark_main
+      benchmark
+      Shlwapi.lib
+  )
   else()
-	  target_link_libraries(arrow_benchmark_main
+    target_link_libraries(arrow_benchmark_main
```
Comment (xhochy) [inline cpp/src/arrow/util/CMakeLists.txt]: Can you remove these reindentations?

**Review by xhochy — APPROVED** [review]
+1, thank you for taking a look at this. There are two code formatting issues reported by Travis. Can you fix these two?

**Comment by seibs** [comment]
Thanks for taking a look @xhochy.

My latest push fixes the formatting issues and also adds a test+fix for a few corner cases my first upload didn't handle correctly.  

**Benchmark with recent changes**
```
Run on (4 X 2208 MHz CPU s)
12/22/17 15:14:31
Benchmark                                        Time           CPU Iterations
------------------------------------------------------------------------------
BM_CopyBitmap/97.6563k/0/min_time:1.000          5 us          5 us     270310   17.9019GB/s
BM_CopyBitmap/976.563k/0/min_time:1.000         65 us         65 us      21854   14.3143GB/s
BM_CopyBitmap/97.6563k/4/min_time:1.000        153 us        152 us       8960   628.592MB/s
BM_CopyBitmap/976.563k/4/min_time:1.000       1534 us       1535 us        896   621.449MB/s
```

w.r.t. the comments above in `CMakeLists.txt`, I actually have a separate pull request open for them at #1406.

**Comment by xhochy** [comment]
@seibs Can resolve the two comments in the other pull request? I would then merge both :)

**Review by wesm — APPROVED** [review]
+1, rebased. Will merge once build passes

**Comment by wesm** [comment]
Rebased again to see if build passes...

**Comment by siddharthteotia** [comment]
What are the general performance concerns we are addressing here? I am just trying to understand if there is something similar to be done on JAVA side for bitvector.

File: cpp/src/arrow/util/bit-util.cc [hunk 3]
```diff
@@ -109,9 +109,37 @@ Status CopyBitmap(MemoryPool* pool, const uint8_t* data, int64_t offset, int64_t
...
+    auto carry_shift = 8U - bit_offset;
+
+    auto carry = 0U;
+    if (BitUtil::BytesForBits(length + bit_offset) > num_bytes) {
+      carry = (data[byte_offset + num_bytes] & carry_mask) << carry_shift;
+    }
+
+    auto i = num_bytes - 1;
+    while (i + 1 > 0) {
+      auto cur_byte = data[byte_offset + i];
```
Comment (wesm) [inline cpp/src/arrow/util/bit-util.cc]: We should write down explicit types for these autos

**Comment by wesm** [comment]
@siddharthteotia there's a couple places in the codebase where we copy bits out of a bitmap, e.g. after slicing vectors https://github.com/apache/arrow/blob/2e3832f34a8933c1ef81331e6d1dde5190366852/cpp/src/arrow/ipc/writer.cc#L63. Depending on the slice offset (whether it's a multiple of 8) we have to either copy 1 bit at a time or do a memcpy

File: cpp/src/arrow/util/bit-util.cc (hunk 3 above)
Comment (seibs) [reply cpp/src/arrow/util/bit-util.cc]: Replaced the autos with types.

**Review by wesm — APPROVED** [review]
+1. Build failing due to some standard flakiness

**Pull request closed without merging**